src/DefinitionResolver.php

```diff
@@ -479,8 +483,10 @@ public function resolveExpressionNodeToType(Node\Expr $expr): Type
                 if ($expr instanceof Node\Expr\MethodCall) {
                     $fqn .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
+                if ($def = $this->index->getDefinition($fqn)) {
+                    if ($def->type instanceof Types\This || $def->type instanceof Types\Self_) {
+                        return $this->resolveExpressionNodeToType($expr->var);
+                    }
```

Comment:
Your example answers my question about `Types\Self`, but it doesn't answer my first question (btw, the correct return type for `method()` would be `@return $this`)